zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs

```diff
@@ -0,0 +1,229 @@
+            // finished the copy yet
+            rw_counter: Delta(3.expr()), // 3 stack pop
+            //program_counter: Delta(1.expr()),
+            stack_pointer: Delta(3.expr()),
+            memory_size: To(memory_expansion.next_memory_size()),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            Some(memory_expansion.gas_cost()),
+        );
+
+        Self {
+            same_context,
+            memory_offset: bytes_memory_offset,
+            data_offset: bytes_data_offset,
+            length: bytes_length,
+            tx_id,
+            calldata_length,
+            memory_expansion,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        tx: &Transaction<F>,
+        _: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let [memory_offset, data_offset, length] =
+            [step.rw_indices[0], step.rw_indices[1], step.rw_indices[2]]
+            .map(|idx| block.rws[idx].stack_value());
+        self.memory_offset.assign(
+            region,
+            offset,
+            Some(memory_offset.to_le_bytes()[..5].try_into().unwrap()),
+        )?;
+        self.data_offset.assign(
+            region,
+            offset,
+            Some(data_offset.to_le_bytes()[..5].try_into().unwrap()),
+        )?;
+        self.length.assign(
+            region,
+            offset,
+            Some(length.to_le_bytes()[..4].try_into().unwrap()),
+        )?;
+        self.tx_id
+            .assign(region, offset, Some(F::from(tx.id as u64)))?;
+        self.calldata_length.assign(
+            region,
+            offset,
+            Some(F::from(tx.call_data_length as u64)),
+        )?;
+
+        // Memory expansion
+        self.memory_expansion.assign(
+            region,
+            offset,
+            step.memory_size,
+            memory_offset.as_u64() + length.as_u64(),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::{
+        step::ExecutionState,
+        test::{rand_word, run_test_circuit_incomplete_fixed_table},
+        util::RandomLinearCombination,
+        witness::{
+            Block, Bytecode, Call, ExecStep, Rw, Transaction,
+        },
+    };
+    use bus_mapping::{
+        bytecode,
+        eth_types::{ToBigEndian, ToLittleEndian, Word},
+        evm::OpcodeId,
+    };
+    use halo2::arithmetic::BaseExt;
+    use pairing::bn256::Fr as Fp;
+
+    fn test_ok(
+        memory_offset: Word,
+        data_offset: Word,
+        length: Word,
+    ) {
+        let bytecode = bytecode! {
+            #[start]
+            .calldatacopy(memory_offset, data_offset, length)
+            STOP
+        };
```

Comment:
we can also do 
```
        let bytecode = bytecode! {
            PUSH32(length)
            PUSH32(data_offset)
            PUSH32(memory_offset)
            #[start]
            CALLDATACOPY
            STOP
        };
```


so that we don't need to add codes to bus-mapping/src/bytecode.rs